src/tools/rbd_wnbd/rbd_wnbd.cc

```diff
@@ -0,0 +1,1282 @@
+        elapsed_ms.QuadPart *= 1000;
+        elapsed_ms.QuadPart /= counter_freq.QuadPart;
+
+        int64_t time_left_ms = max((int64_t)0, timeout_ms - elapsed_ms.QuadPart);
+
+        cfg.hard_disconnect = hard_disconnect || !time_left_ms;
+        cfg.hard_disconnect_fallback = true;
+        cfg.soft_disconnect_timeout = time_left_ms / 1000;
+
+        dout(5) << "Removing mapping: " << cfg.devpath
+                << ". Timeout: " << cfg.soft_disconnect_timeout
+                << "ms. Hard disconnect: " << cfg.hard_disconnect
+                << dendl;
+
+        r = do_unmap(&cfg, unregister);
+        if (r) {
+          err = r;
+          derr << "Could not remove mapping: " << cfg.devpath
+               << ". Error: " << r << dendl;
+        }
+        else {
+          dout(5) << "Successfully removed mapping: " << cfg.devpath << dendl;
+        }
+      });
+  }
+  pool.join();
+
+  r = iterator.get_error();
+  if (r) {
+    derr << "Could not fetch all mappings. Error: " << r << dendl;
+    err = r;
+  }
+
+  return err;
+}
+
+class RBDService : public ServiceBase {
+  private:
+    bool hard_disconnect;
+    int soft_disconnect_timeout;
+    int thread_count;
+
+  public:
+    RBDService(bool _hard_disconnect,
+               int _soft_disconnect_timeout,
+               int _thread_count)
+      : ServiceBase(g_ceph_context)
+      , hard_disconnect(_hard_disconnect)
+      , soft_disconnect_timeout(_soft_disconnect_timeout)
+      , thread_count(_thread_count)
+    {
+    }
+
+    int run_hook() override {
+      return restart_registered_mappings(thread_count);
+    }
+    // Invoked when the service is requested to stop.
+    int stop_hook() override {
+      return disconnect_all_mappings(
+        false, hard_disconnect, soft_disconnect_timeout, thread_count);
+    }
+    // Invoked when the system is shutting down.
+    int shutdown_hook() override {
+      return stop_hook();
+    }
+};
+
+static void usage()
+{
+  const char* usage_str =R"(
+Usage: rbd-wnbd [options] map <image-or-snap-spec>           Map an image to wnbd device
+                [options] unmap <device|image-or-snap-spec>  Unmap wnbd device
+                [options] list                               List mapped wnbd devices
+                [options] show-mapped <image-or-snap-spec>   Show mapped wnbd device
+                stats <image-or-snap-spec>                   Show IO counters
+                [options] service                            Windows service entrypoint,
+                                                             handling device lifecycle
+
+Map options:
+  --device <device path>  Optional mapping unique identifier
+  --exclusive             Forbid writes by other clients
+  --read-only             Map read-only
+  --io-req-workers        The number of workers that dispatch IO requests.
+                          Default: 4
+  --io-reply-workers      The number of workers that dispatch IO replies.
```

Comment:
What's the purpose of this ASIO thread pool? librbd will be dispatching from its own ASIO thread pool (defaults to 2 threads) which then just context-switches to this one?

Reply:
In order to make some optimizations, it's important for us to know which threads are going to send the replies. By looking at the librbd code, I think the async IO callbacks can be executed outside the ASIO thread pool.

Windows async IO operations use ``OVERLAPPED`` [1] structures, having an embedded event that signals the completion of the async operation. I noticed that initializing those events/structures can be quite time consuming in some cases, so we're trying to reuse them, using TLS to retrieve cached ``OVERLAPPED`` structures.

[1] https://docs.microsoft.com/en-us/windows/win32/api/minwinbase/ns-minwinbase-overlapped